File: algorithms-well-known/union.go
package main

import "fmt"

type Union struct {
	Parent []int
	Size []int
}

func (un *Union) Root(p int) int {
	for p != un.Parent[p] {
		un.Parent[p] = un.Parent[un.Parent[p]]
		p = un.Parent[p]
	}
	return p
}

func (un *Union) Connected(p int, q int) bool {
	return un.Root(p) == un.Root(q)
}

func (un *Union) Union(p int, q int) {
	root_p, root_q := un.Root(p), un.Root(q)
	if root_p != root_q {
		if un.Size[root_p] < un.Size[root_q] {
			un.Parent[root_p] = root_q
			un.Size[root_q] += un.Size[root_p]
		} else {
			un.Parent[root_p] = root_q
			un.Size[root_q] += un.Size[root_p]
		}
	}
}

func NewUnion(n int) Union {
	un := Union{make([]int, 0), make([]int, 0)}
	for i := 0; i < n; i++ {
		un.Parent = append(un.Parent, i)
		un.Size = append(un.Size, 1)
	}
	return un
}

File: alortihms-before-interview/graph/1_union_find/_union.py
# 1. Implement root(), connected() and union()
#
# 2. Time Complexity:
#
#   Appending makes the depth of tree to log(n).
#   Path compression makes the running time to c(N+M log*n).
#
# --END--


class Union:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.size = [1] * n

    def root(self, p):
        while p != self.parent[p]:
            # path compression
            self.parent[p] = self.parent[self.parent[p]]
            p = self.parent[p]
        return p

    def connected(self, p, q):
        return self.root(p) == self.root(q)

    def union(self, p, q):
        # smaller tree to bigger tree
        root_p, root_q = self.root(p), self.root(q)
        if root_p != root_q:
            if self.size[root_p] < self.size[root_q]:
                self.parent[root_p] = root_q
                self.size[root_q] += self.size[root_p]
            else:
                self.parent[root_q] = root_p
                self.size[root_p] += self.size[root_q]
